parquet/src/main/java/org/apache/iceberg/parquet/ParquetSchemaUtil.java

```diff
@@ -83,33 +92,144 @@ public static MessageType pruneColumnsFallback(MessageType fileSchema, Schema ex
   }
 
   public static boolean hasIds(MessageType fileSchema) {
-    try {
-      // Try to convert the type to Iceberg. If an ID assignment is needed, return false.
-      ParquetTypeVisitor.visit(fileSchema, new MessageTypeToType(fileSchema) {
-        @Override
-        protected int nextId() {
-          throw new IllegalStateException("Needed to assign ID");
+    return ParquetTypeVisitor.visit(fileSchema, new HasIds(), true);
+  }
+
+  public static MessageType addFallbackIds(MessageType fileSchema, NameMapping nameMapping) {
+    if (nameMapping == null) {
+      MessageTypeBuilder builder = org.apache.parquet.schema.Types.buildMessage();
+
+      int ordinal = 1; // ids are assigned starting at 1
+      for (Type type : fileSchema.getFields()) {
+        builder.addField(type.withId(ordinal));
+        ordinal += 1;
+      }
+
+      return builder.named(fileSchema.getName());
+    } else {
+      return (MessageType) ParquetTypeVisitor.visit(fileSchema, new AssignIdsByNameMapping(nameMapping), true);
+    }
+  }
+
+  public static class HasIds extends ParquetTypeVisitor<Boolean> {
+    @Override
+    public Boolean message(MessageType message, List<Boolean> fields) {
+      return struct(message, fields);
+    }
+
+    @Override
+    public Boolean struct(GroupType struct, List<Boolean> hasIds) {
+      for (Boolean hasId : hasIds) {
+        if (hasId) {
+          return true;
         }
-      });
+      }
+      return struct.getId() != null;
+    }
+
+    @Override
+    public Boolean list(GroupType array, Boolean hasId) {
+      if (hasId) {
+        return true;
+      } else {
+        return array.getId() != null;
+      }
+    }
 
-      // no assignment was needed
-      return true;
+    @Override
+    public Boolean map(GroupType map, Boolean keyHasId, Boolean valueHasId) {
+      if (keyHasId || valueHasId) {
+        return true;
+      } else {
+        return map.getId() != null;
+      }
+    }
 
-    } catch (IllegalStateException e) {
-      // at least one field was missing an id.
-      return false;
+    @Override
+    public Boolean primitive(PrimitiveType primitive) {
+      return primitive.getId() != null;
     }
   }
 
-  public static MessageType addFallbackIds(MessageType fileSchema) {
-    MessageTypeBuilder builder = org.apache.parquet.schema.Types.buildMessage();
+  public static class AssignIdsByNameMapping extends ParquetTypeVisitor<Type> {
```

Comment:
This overwrites any existing IDs, so let's call it `applyNameMapping` instead.

Also, this is quite a large class. Can you move it into its own file? It should be package-private.

Reply:
Done.